go/types/type_cache.go

```diff
@@ -537,6 +537,10 @@ func MakeUnionType(elemTypes ...*Type) *Type {
 	return makeSimplifiedUnion(elemTypes...)
 }
 
+func MakeFullUnionType(elemTypes ...*Type) *Type {
```

Comment:
I feel like the terminology is getting confused here. Let's ask @arv in meatspace what the right terms are.

This thing you're adding now feels more like a _union_ of the input types than what I had. The thing I had might be called `ExtractCommonSupertype()` ?